From 91c036eed7ddd62e0029a5fd5066b5ae8a20f09d Mon Sep 17 00:00:00 2001
From: Igal Zeltser <izeltser@habana.ai>
Date: Sun, 20 Nov 2022 11:17:25 +0200
Subject: [PATCH] [SW-66410] gaudi3: initialize pcie drain

Pcie drain is configured and enabled, perliminary timeout values are
taken from the LKD.

Change-Id: I02a7918184408cde628be3f209c3dcdeddd231c7
Signed-off-by: Igal Zeltser <izeltser@habana.ai>
---

diff --git a/arch/gaudi3/arc/common.c b/arch/gaudi3/arc/common.c
index 4d39023..de8fa44 100644
--- a/arch/gaudi3/arc/common.c
+++ b/arch/gaudi3/arc/common.c
@@ -31,6 +31,8 @@
 #include <d0_parc_global_conf_regs.h>
 #include <d0_parc_global_conf_masks.h>
 #include <d0_psoc_global_conf_masks.h>
+#include <d0_pcie_wrap_regs.h>
+#include <d0_pcie_wrap_masks.h>
 #include <d1_pcie_aux_regs.h>
 #include <d1_pcie_core_regs.h>
 #include <d1_pcie_dbi_sig_regs.h>
@@ -184,6 +186,32 @@
 	regmap_writel(PSOC_ARC2_PRELOADED_ID, mmD0_PARC_ARC2_CFG_ARC_NUM);
 }
 
+#define HBW_DRAIN_TIMEOUT	0x1000
+#define LBW_DRAIN_TIMEOUT	0xF000
+
+void ppboot_pcie_drain_init(void)
+{
+	u32 val;
+
+	/* If FW skip enabled return */
+	if (hl_skip_fw_cfg())
+		return;
+
+	regmap_writel(HBW_DRAIN_TIMEOUT, mmD0_PCIE_WRAP_HBW_DRAIN_TIMEOUT);
+	regmap_writel(D0_PCIE_WRAP_HBW_RSP_ERR_DRAIN_STAMP_EN_M,
+		      mmD0_PCIE_WRAP_HBW_RSP_ERR_DRAIN_STAMP);
+	val = regmap_readl(mmD0_PCIE_WRAP_HBW_DRAIN_CFG);
+	val |= D0_PCIE_WRAP_HBW_DRAIN_CFG_EN_M;
+	regmap_writel(val, mmD0_PCIE_WRAP_HBW_DRAIN_CFG);
+
+	regmap_writel(LBW_DRAIN_TIMEOUT, mmD0_PCIE_WRAP_LBW_DRAIN_TIMEOUT);
+	regmap_writel(D0_PCIE_WRAP_LBW_RSP_ERR_DRAIN_STAMP_EN_M,
+		      mmD0_PCIE_WRAP_LBW_RSP_ERR_DRAIN_STAMP);
+	val = regmap_readl(mmD0_PCIE_WRAP_LBW_DRAIN_CFG);
+	val |= D0_PCIE_WRAP_LBW_DRAIN_CFG_EN_M;
+	regmap_writel(val, mmD0_PCIE_WRAP_LBW_DRAIN_CFG);
+}
+
 void hl_plat_early_init(void)
 {
 	hl_lbw_early_write_resp_enable();
